build,deps,tools: prepare to update to OpenSSL 3.5

richardlau opened the pull request:
- **test: prepare test-crypto-rsa-dsa for newer OpenSSL**
Update `parallel/test-crypto-rsa-dsa` to prepare for updating
`deps/openssl` to later versions of OpenSSL which support implicit
rejections with `RSA_PKCS1_PADDING`.
- **build,deps,tools: prepare to update to OpenSSL 3.5**
Update the updater script and files under `deps/openssl/config` in
preparation for updating `deps/openssl` to OpenSSL 3.5.

---

Note that this PR does not do the actual upgrade -- that should be handled by running the workflow after this lands.

This builds upon work started by @targos in https://github.com/nodejs/node/pull/57795.

aduh95 commented:
Can we backport this to LTS? Should we add the labels?

richardlau commented:
> Can we backport this to LTS? Should we add the lts-watch labels?

I've stuck a watch label for Node.js 22 as we have to update that at some point because OpenSSL 3.0 reaches End-of-Life in September 2026 which is before the End-of-Life of Node.js 22 (end of April 2027).

For Node.js 20 we could stay on OpenSSL 3.0 as Node.js 20 will reach End-of-Life at the end of April 2026 which is prior to End-of-Life of OpenSSL 3.0. But we can have a separate discussion about how we would manage updates of OpenSSL across all of our LTS versions.

richardlau commented:
Moving this to draft. I ran a CI on https://github.com/richardlau/node-1/tree/openssl-3.5 (from which the commits in this PR are based) and that has failed to build on Windows: https://ci.nodejs.org/job/node-test-commit/79522/

I suspect we'll need to add even more entries into `deps/openssl/config/Makefile_VC-WIN*` for the missing `*.asm` file(s). (There must be a better way of doing this than manually editing.)

targos commented:
I tried the update script again with the changes from this PR on my fork: https://github.com/targos/node/pull/27
Workflow run: https://github.com/targos/node/actions/runs/15186079568
It does the job fine without https://github.com/nodejs/node/pull/58099

targos commented:
^ Which makes sense now that I think about it again. The "Regenerate platform specific files" step has access to the files even though they are ignored by git.

I'm preparing a PR to restore the necessary files.

targos commented:
https://github.com/nodejs/node/pull/58431

richardlau commented:
I've updated this PR. I've tested it [on my fork](https://github.com/richardlau/node-1/actions/runs/16511765267) which opened a [pull request against my fork updating to OpenSSL 3.5.1](https://github.com/richardlau/node-1/pull/20) that I ran a [CI on which was yellow](https://ci.nodejs.org/job/node-test-commit/81508/) (i.e. good enough to land). For good measure, I'm running through the same testing again now that I've squashed commits together prior to updating this PR.

Again, a reminder that this PR doesn't contain the OpenSSL update -- it updates the scripts so that the [OpenSSL update GitHub workflow](https://github.com/nodejs/node/actions/workflows/update-openssl.yml) will perform the update when run.

The config generation now runs in the existing [container](https://github.com/nodejs/node/blob/main/deps/openssl/config/Dockerfile) which for now will keep us on an older version of GNU Assembler that will result in generated assembler being compatible with the version of GNU assembler currently used in the older Linux distributions we support.

A significant delay in getting this PR working (other than the distraction of my prolonged change of employer) was debugging issues with extending the handcrafted Windows makefiles in `deps/openssl/config`.

(FYI if there are any changes requested/needed, I'm out of office next week and may not be responsive until I'm back.)
